### PR TITLE
[SCEVExpander] Attempt to reinfer flags dropped due to CSE

### PR DESCRIPTION
**Pull request by preames**
LSR uses SCEVExpander to generate induction formulas.  The expander internally tries to reuse existing IR expressions.  To do that, it needs to strip any poison generating flags (nsw, nuw, exact, nneg, etc..) which may not be valid for the newly added users.

This is conservatively correct, but has the effect that LSR will strip nneg flags on zext instructions involved in trip counts in loop preheaders. To avoid this, this patch adjusts the expanded to reinfer the flags on the CSE candidate if legal for all possible users.

This should fix the regression reported in https://github.com/llvm/llvm-project/issues/71200.

This should arguably be done inside canReuseInstruction instead, but doing it outside is more conservative compile time wise.  Both canReuseInstruction and isGuaranteedNotToBePoison walk operand lists, so right now we are performing work which is roughly O(N^2) in the size of the operand graph. We should fix that before making the per operand step more expensive.  My tenative plan is to land this, and then rework the code to sink the logic into more core interfaces.

### CODE REVIEW

**Comment by preames** [comment]
ping

File: llvm/lib/Transforms/Utils/ScalarEvolutionExpander.cpp [hunk 1]
```diff
@@ -1532,8 +1532,28 @@ Value *SCEVExpander::expand(const SCEV *S) {
...
+          BO->setHasNoSignedWrap(
+            ScalarEvolution::maskFlags(*Flags, SCEV::FlagNSW) == SCEV::FlagNSW);
+        }
+      if (auto *NNI = dyn_cast<PossiblyNonNegInst>(I)) {
+        auto *Src = NNI->getOperand(0);
+        if (SE.isKnownNonNegative(SE.getSCEV(Src)) ||
+            isKnownNonNegative(Src, DL, 0, &SE.AC, I, &SE.DT) ||
+            isImpliedByDomCondition(ICmpInst::ICMP_SGE, Src,
+                                    Constant::getNullValue(Src->getType()), I,
+                                    DL).value_or(false))
```
Comment (nikic) [inline llvm/lib/Transforms/Utils/ScalarEvolutionExpander.cpp:1553]: It looks like you don't have test coverage for all three of these calls.

Comment (nikic) [reply llvm/lib/Transforms/Utils/ScalarEvolutionExpander.cpp:1553]: Generally I'd expect ValueTracking `isKnownNonNegative` to not give any useful additional information here, with the exception of `assume` s.

**Review by nikic — COMMENTED** [review]
I think this is generally fine in terms of approach. Surprisingly, it doesn't seem to have any significant impact on compile-time either.

**Review by nikic — APPROVED** [review]
LGTM

**Pull request merged**